File: packages/py/tracing/src/uselemma_tracing/run_batch_span_processor.py
from __future__ import annotations

import threading
import uuid
from collections.abc import Mapping
from typing import Any

from opentelemetry.context import Context
from opentelemetry.sdk.trace import ReadableSpan, Span
from opentelemetry.sdk.trace.export import SpanExporter, SpanProcessor


class RunBatchSpanProcessor(SpanProcessor):
    def __init__(self, exporter: SpanExporter) -> None:
        self._exporter = exporter
        self._lock = threading.Lock()
        self._shutdown = False
        self._span_id_to_run_id: dict[int, str] = {}
        self._top_level_span_id_by_run_id: dict[str, int] = {}
        self._direct_child_count_by_run_id: dict[str, int] = {}
        self._direct_child_span_id_to_run_id: dict[int, str] = {}
        self._batches: dict[str, list[ReadableSpan]] = {}
        self._ended_runs: set[str] = set()

    def on_start(self, span: Span, parent_context: Context | None = None) -> None:
        span_id = span.context.span_id

        if self._is_top_level_run(span):
            run_id = self._get_attr(span, "lemma.run_id") or str(uuid.uuid4())
            span.set_attribute("lemma.run_id", run_id)
            with self._lock:
                self._span_id_to_run_id[span_id] = run_id
                self._top_level_span_id_by_run_id[run_id] = span_id
                self._direct_child_count_by_run_id[run_id] = 0
            return

        parent = getattr(span, "parent", None)
        if parent is None:
            return

        with self._lock:
            run_id = self._span_id_to_run_id.get(parent.span_id)
            if run_id is None:
                return

            if self._top_level_span_id_by_run_id.get(run_id) == parent.span_id:
                self._direct_child_span_id_to_run_id[span_id] = run_id
                self._direct_child_count_by_run_id[run_id] = (
                    self._direct_child_count_by_run_id.get(run_id, 0) + 1
                )

            self._span_id_to_run_id[span_id] = run_id

        span.set_attribute("lemma.run_id", run_id)

    def on_end(self, span: ReadableSpan) -> None:
        span_id = span.context.span_id
        parent = getattr(span, "parent", None)

        with self._lock:
            run_id = self._span_id_to_run_id.get(span_id)
            if run_id is None and parent is not None:
                run_id = self._span_id_to_run_id.get(parent.span_id)

            if run_id is None:
                return

            is_top_level_run = self._is_top_level_run(span)
            should_skip_export = self._should_skip_export(span)
            direct_child_run_id = self._direct_child_span_id_to_run_id.pop(span_id, None)
            if direct_child_run_id is not None:
                self._direct_child_count_by_run_id[direct_child_run_id] = max(
                    0, self._direct_child_count_by_run_id.get(direct_child_run_id, 0) - 1
                )

            if not should_skip_export:
                self._batches.setdefault(run_id, []).append(span)

            if is_top_level_run:
                self._ended_runs.add(run_id)

        self._export_run_batch(run_id, force=False)

    def shutdown(self) -> None:
        with self._lock:
            if self._shutdown:
                return

            self._shutdown = True

        self.force_flush()
        self._exporter.shutdown()

    def force_flush(self, timeout_millis: int = 30000) -> bool:
        with self._lock:
            run_ids = list(self._batches.keys())

        for run_id in run_ids:
            self._export_run_batch(run_id, force=True)

        flush_result = self._exporter.force_flush(timeout_millis=timeout_millis)
        return True if flush_result is None else bool(flush_result)

    def _export_run_batch(self, run_id: str, force: bool) -> None:
        with self._lock:
            if not force and (
                run_id not in self._ended_runs
                or not self._has_no_open_direct_children_locked(run_id)
            ):
                return

            batch = self._batches.pop(run_id, [])
            self._ended_runs.discard(run_id)
            self._clear_run_mapping_locked(run_id)

        if not batch:
            return

        self._exporter.export(batch)

    def _clear_run_mapping_locked(self, run_id: str) -> None:
        stale_span_ids = [
            span_id
            for span_id, mapped_run_id in self._span_id_to_run_id.items()
            if mapped_run_id == run_id
        ]
        for span_id in stale_span_ids:
            del self._span_id_to_run_id[span_id]
        stale_direct_child_span_ids = [
            span_id
            for span_id, mapped_run_id in self._direct_child_span_id_to_run_id.items()
            if mapped_run_id == run_id
        ]
        for span_id in stale_direct_child_span_ids:
            del self._direct_child_span_id_to_run_id[span_id]
        self._top_level_span_id_by_run_id.pop(run_id, None)
        self._direct_child_count_by_run_id.pop(run_id, None)

    def _has_no_open_direct_children_locked(self, run_id: str) -> bool:
        return self._direct_child_count_by_run_id.get(run_id, 0) == 0

    @staticmethod
    def _is_top_level_run(span: Span | ReadableSpan) -> bool:
        return span.name == "ai.agent.run" and getattr(span, "parent", None) is None

    @staticmethod
    def _scope_name(span: Span | ReadableSpan) -> str | None:
        scope = getattr(span, "instrumentation_scope", None)
        return getattr(scope, "name", None)

    @classmethod
    def _should_skip_export(cls, span: Span | ReadableSpan) -> bool:
        return cls._scope_name(span) == "next.js"

    @staticmethod
    def _get_attr(span: Span, key: str) -> str | None:
        attributes = getattr(span, "attributes", None)
        if not isinstance(attributes, Mapping):
            return None
        value: Any = attributes.get(key)
        return value if isinstance(value, str) and value else None

File: packages/ts/tracing/src/run-batch-span-processor.ts
import { randomUUID } from "crypto";
import type { Context } from "@opentelemetry/api";
import {
  type ReadableSpan,
  type Span,
  type SpanExporter,
  type SpanProcessor,
} from "@opentelemetry/sdk-trace-base";

type RunId = string;
type SpanId = string;

export class RunBatchSpanProcessor implements SpanProcessor {
  private isShutdown = false;
  private spanIdToRunId = new Map<SpanId, RunId>();
  private topLevelSpanIdByRunId = new Map<RunId, SpanId>();
  private directChildCountByRunId = new Map<RunId, number>();
  private directChildSpanIdToRunId = new Map<SpanId, RunId>();
  private batches = new Map<RunId, ReadableSpan[]>();
  private endedRuns = new Set<RunId>();
  private readonly exporter: SpanExporter;

  constructor(exporter: SpanExporter) {
    this.exporter = exporter;
  }

  onStart(span: Span, _parentContext: Context): void {
    const spanId = span.spanContext().spanId;

    if (this.isTopLevelRun(span)) {
      const runId = this.getRunIdFromSpan(span) ?? randomUUID();
      span.setAttribute("lemma.run_id", runId);
      this.spanIdToRunId.set(spanId, runId);
      this.topLevelSpanIdByRunId.set(runId, spanId);
      this.directChildCountByRunId.set(runId, 0);
      return;
    }

    const parentSpanId = span.parentSpanContext?.spanId;
    if (!parentSpanId) return;

    const runId = this.spanIdToRunId.get(parentSpanId);
    if (!runId) return;

    if (this.topLevelSpanIdByRunId.get(runId) === parentSpanId) {
      this.directChildSpanIdToRunId.set(spanId, runId);
      this.directChildCountByRunId.set(runId, (this.directChildCountByRunId.get(runId) ?? 0) + 1);
    }

    this.spanIdToRunId.set(spanId, runId);
    span.setAttribute("lemma.run_id", runId);
  }

  onEnd(span: ReadableSpan): void {
    const spanId = span.spanContext().spanId;
    const runId =
      this.spanIdToRunId.get(spanId) ??
      (span.parentSpanContext?.spanId
        ? this.spanIdToRunId.get(span.parentSpanContext.spanId)
        : undefined);

    if (!runId) return;

    const isTopLevelRun = this.isTopLevelRun(span);
    const shouldSkipExport = this.shouldSkipExport(span);
    const directChildRunId = this.directChildSpanIdToRunId.get(spanId);
    if (directChildRunId) {
      this.directChildSpanIdToRunId.delete(spanId);
      const currentCount = this.directChildCountByRunId.get(directChildRunId) ?? 0;
      this.directChildCountByRunId.set(directChildRunId, Math.max(0, currentCount - 1));
    }

    if (!shouldSkipExport) {
      const batch = this.batches.get(runId);
      if (batch) batch.push(span);
      else this.batches.set(runId, [span]);
    }

    if (isTopLevelRun) {
      this.endedRuns.add(runId);
    }

    void this.exportRunBatch(runId, false);
  }

  async forceFlush(): Promise<void> {
    const runIds = [...this.batches.keys()];
    await Promise.all(runIds.map((runId) => this.exportRunBatch(runId, true)));
    if (this.exporter.forceFlush) {
      await this.exporter.forceFlush();
    }
  }

  async shutdown(): Promise<void> {
    if (this.isShutdown) return;

    this.isShutdown = true;
    await this.forceFlush();
    this.exporter.shutdown();
  }

  private isTopLevelRun(span: Span | ReadableSpan): boolean {
    return span.name === "ai.agent.run" && !span.parentSpanContext;
  }

  private getRunIdFromSpan(span: Span): string | undefined {
    const attributes = (span as unknown as { attributes?: Record<string, unknown> })
      .attributes;
    const runId = attributes?.["lemma.run_id"];
    return typeof runId === "string" && runId.length > 0 ? runId : undefined;
  }

  private getInstrumentationScopeName(span: Span | ReadableSpan): string | undefined {
    return (
      span as unknown as {
        instrumentationScope?: { name?: string };
      }
    ).instrumentationScope?.name;
  }

  private shouldSkipExport(span: Span | ReadableSpan): boolean {
    return this.getInstrumentationScopeName(span) === "next.js";
  }

  private async exportRunBatch(runId: RunId, force: boolean): Promise<void> {
    if (!force && (!this.endedRuns.has(runId) || !this.hasNoOpenDirectChildren(runId))) {
      return;
    }

    const batch = this.batches.get(runId);

    this.batches.delete(runId);
    this.endedRuns.delete(runId);
    this.clearRunMapping(runId);

    if (!batch || batch.length === 0) return;

    await new Promise<void>((resolve) => {
      this.exporter.export(batch, () => resolve());
    });
  }

  private hasNoOpenDirectChildren(runId: RunId): boolean {
    return (this.directChildCountByRunId.get(runId) ?? 0) === 0;
  }

  private clearRunMapping(runId: RunId): void {
    for (const [spanId, mappedRunId] of this.spanIdToRunId.entries()) {
      if (mappedRunId === runId) {
        this.spanIdToRunId.delete(spanId);
      }
    }
    for (const [spanId, mappedRunId] of this.directChildSpanIdToRunId.entries()) {
      if (mappedRunId === runId) {
        this.directChildSpanIdToRunId.delete(spanId);
      }
    }
    this.topLevelSpanIdByRunId.delete(runId);
    this.directChildCountByRunId.delete(runId);
  }
}
